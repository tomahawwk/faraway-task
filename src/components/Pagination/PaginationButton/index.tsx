import {FC} from 'react';

interface IPaginationButton {
  active: boolean;
  number: number;
  onClick(): void;
}

const PaginationButton: FC<IPaginationButton> = ({active, number, onClick}) => {
  return (
    <button
      className={`hidden md:block rounded-xs px-xs py-[5px] text-[14px] duration-300 will-change-transform
        ${
          active
            ? 'bg-grey pointer-events-none'
            : 'bg-secondary-dark hover:bg-secondary-main hover:scale-[1.1]'
        }
        `}
      onClick={onClick}>
      {number}
    </button>
  );
};

export default PaginationButton;
