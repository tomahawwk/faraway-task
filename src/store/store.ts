import {combineReducers, configureStore} from '@reduxjs/toolkit';
import charactersReducer from './reducers/CharactersSlice';
import characterReducer from './reducers/CharacterSlice';
import filtersReducer from './reducers/FiltersSlice';

const rootReducer = combineReducers({
  charactersReducer,
  characterReducer,
  filtersReducer,
});

export const setupStore = () => {
  return configureStore({
    reducer: rootReducer,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        serializableCheck: false,
      }),
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
