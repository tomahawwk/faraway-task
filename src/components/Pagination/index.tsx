import {useState, FC, useEffect} from 'react';
import {useAppDispatch, useAppSelector} from 'store/hooks/redux';
import {
  getCharactersSelector,
  setCurrentPageIndex,
} from 'store/reducers/CharactersSlice';
import PaginationButton from './PaginationButton';

interface PaginationProps {
  next: string | null;
  prev: string | null;
  count: number;
  onPageChange: (pageIndex: number) => void;
}

const Pagination: FC<PaginationProps> = ({next, prev, count, onPageChange}) => {
  const {currentPageIndex} = useAppSelector(getCharactersSelector);
  const [currentPage, setCurrentPage] = useState<number>(currentPageIndex);
  const dispatch = useAppDispatch();
  const totalPages = Math.ceil(count / 10);

  useEffect(() => {
    if (next) {
      const match = next.match(/page=(\d+)/);
      if (match) {
        setCurrentPage(parseInt(match[1], 10) - 1);
        dispatch(setCurrentPageIndex(parseInt(match[1], 10) - 1));
      }
    } else if (prev) {
      const match = prev.match(/page=(\d+)/);
      if (match) {
        setCurrentPage(parseInt(match[1], 10) + 1);
        dispatch(setCurrentPageIndex(parseInt(match[1], 10) + 1));
      }
    }
  }, [next, prev]);

  const handlePageChange = (pageIndex: number) => {
    setCurrentPage(pageIndex);
    onPageChange(pageIndex);
    dispatch(setCurrentPageIndex(pageIndex));
  };

  const renderPageButtons = () => {
    const maxButtons = 5;
    const visiblePages = [];

    let startPage = Math.max(1, currentPage - Math.floor(maxButtons / 2));
    let endPage = Math.min(
      totalPages,
      currentPage + Math.floor(maxButtons / 2),
    );

    if (endPage - startPage + 1 > maxButtons) {
      if (currentPage > Math.floor(maxButtons / 2)) {
        startPage = currentPage - Math.floor(maxButtons / 2) + 1;
        endPage = startPage + maxButtons - 1;
      } else {
        endPage = Math.min(totalPages, maxButtons);
        startPage = 1;
      }
    }

    visiblePages.push(
      <PaginationButton
        key={1}
        active={1 === currentPage}
        number={1}
        onClick={() => handlePageChange(1)}
      />,
    );

    if (startPage > 2) {
      visiblePages.push(
        <span key="dots1" className="px-xs ml-1 hidden md:block">
          ...
        </span>,
      );
    }

    for (let i = startPage; i <= endPage; i++) {
      if (i !== 1 && i !== totalPages) {
        visiblePages.push(
          <PaginationButton
            key={i}
            active={i === currentPage}
            number={i}
            onClick={() => handlePageChange(i)}
          />,
        );
      }
    }

    if (endPage < totalPages - 1) {
      visiblePages.push(
        <span key="dots2" className="px-xs hidden md:block">
          ...
        </span>,
      );
    }

    visiblePages.push(
      <PaginationButton
        key={totalPages}
        active={totalPages === currentPage}
        number={totalPages}
        onClick={() => handlePageChange(totalPages)}
      />,
    );

    return visiblePages;
  };

  return (
    <div className="flex justify-center items-center gap-[6px]">
      <button
        className="w-[30px]"
        disabled={!prev}
        onClick={() => handlePageChange(currentPage - 1)}>
        <img src="images/prev-arrow.svg" className="w-full" alt="previous" />
      </button>
      {renderPageButtons()}
      <button
        className="w-[30px]"
        disabled={!next}
        onClick={() => handlePageChange(currentPage + 1)}>
        <img src="images/next-arrow.svg" className="w-full" alt="next" />
      </button>
    </div>
  );
};

export default Pagination;
