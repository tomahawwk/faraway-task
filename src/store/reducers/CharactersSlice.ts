import {PayloadAction, createSlice} from '@reduxjs/toolkit';
import {fetchCharacters} from './ActionCreators';
import {ICharacter} from 'store/models/ICharacter';
import {RootState} from '../store';

interface CharactersState {
  characters: ICharacter[];
  isLoading: boolean;
  next: string;
  prev: string;
  count: number;
  currentPageIndex: number;
}

const initialState: CharactersState = {
  characters: [],
  next: '',
  prev: '',
  count: 0,
  isLoading: false,
  currentPageIndex: 1,
};

export const charactersSlice = createSlice({
  name: 'characters',
  initialState,
  reducers: {
    setCurrentPageIndex(state, action: PayloadAction<number>) {
      state.currentPageIndex = action.payload;
    },
  },
  extraReducers: builder => {
    builder
      .addCase(
        fetchCharacters.fulfilled.type,
        (state, action: PayloadAction<any>) => {
          state.isLoading = false;
          state.characters = action.payload.results;
          state.next = action.payload.next;
          state.prev = action.payload.previous;
          state.count = action.payload.count;
        },
      )
      .addCase(fetchCharacters.pending.type, state => {
        state.isLoading = true;
      });
  },
});

export default charactersSlice.reducer;

export const getCharactersSelector = (state: RootState) =>
  state.charactersReducer;

export const {setCurrentPageIndex} = charactersSlice.actions;
