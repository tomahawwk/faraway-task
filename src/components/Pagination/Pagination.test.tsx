import {render, screen, fireEvent} from '@testing-library/react';
import {Provider} from 'react-redux';
import {configureStore} from '@reduxjs/toolkit';
import CharactersSlice from 'store/reducers/CharactersSlice';
import PaginationButton from './PaginationButton'; // Добавьте импорт компонента PaginationButton
import {MemoryRouter} from 'react-router-dom';
import Pagination from '.';
import {setupStore} from 'store/store';

jest.mock('./PaginationButton', () => ({
  __esModule: true,
  default: jest.fn(({number, active, onClick}) => (
    <button
      data-testid={`pagination-button-${number}`}
      className={`pagination-button ${active ? 'active' : ''}`}
      onClick={onClick}>
      {number}
    </button>
  )),
}));

const store = setupStore();

describe('Pagination component', () => {
  const mockProps = {
    next: 'https://swapi.dev/api/character?page=2',
    prev: null,
    count: 20, // Количество элементов для пагинации
    onPageChange: jest.fn(), // Моковая функция для обработки изменения страницы
  };

  test('renders pagination buttons correctly', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Pagination {...mockProps} />
        </MemoryRouter>
      </Provider>,
    );
    setTimeout(() => {
      expect(screen.getByTestId('pagination-button-2')).toBeInTheDocument();
      expect(screen.getByTestId('pagination-button-2')).toBeInTheDocument();
      expect(screen.getByTestId('pagination-button-3')).toBeInTheDocument();
      expect(screen.getByTestId('pagination-button-4')).toBeInTheDocument();
      expect(screen.getByTestId('pagination-button-5')).toBeInTheDocument();
    }, 10);
  });

  test('disables previous button when there is no previous page', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Pagination
            {...mockProps}
            next="https://swapi.dev/api/character?page=2"
            prev={null}
          />
        </MemoryRouter>
      </Provider>,
    );

    const prevButton = screen.getByRole('button', {name: /previous/i});
    expect(prevButton).toBeDisabled();
  });

  test('disables next button when there is no next page', () => {
    render(
      <Provider store={store}>
        <MemoryRouter>
          <Pagination
            {...mockProps}
            next={null}
            prev="https://swapi.dev/api/character?page=1"
          />
        </MemoryRouter>
      </Provider>,
    );

    const nextButton = screen.getByRole('button', {name: /next/i});
    expect(nextButton).toBeDisabled();
  });

  test('Pagination snapshot', () => {
    const pagination = render(
      <Provider store={store}>
        <MemoryRouter>
          <Pagination
            {...mockProps}
            next={null}
            prev="https://swapi.dev/api/character?page=1"
          />
        </MemoryRouter>
      </Provider>,
    );
    expect(pagination).toMatchSnapshot();
  });
});
