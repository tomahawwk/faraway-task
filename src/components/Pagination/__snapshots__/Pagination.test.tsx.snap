// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination component Pagination snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="flex justify-center items-center gap-[6px]"
      >
        <button
          class="w-[30px]"
        >
          <img
            alt="previous"
            class="w-full"
            src="images/prev-arrow.svg"
          />
        </button>
        <button
          class="w-[30px]"
          disabled=""
        >
          <img
            alt="next"
            class="w-full"
            src="images/next-arrow.svg"
          />
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="flex justify-center items-center gap-[6px]"
    >
      <button
        class="w-[30px]"
      >
        <img
          alt="previous"
          class="w-full"
          src="images/prev-arrow.svg"
        />
      </button>
      <button
        class="w-[30px]"
        disabled=""
      >
        <img
          alt="next"
          class="w-full"
          src="images/next-arrow.svg"
        />
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
